File: elements-of-programming-interviews/python/parity.py
from test_framework import generic_test
from typing import List


def parity(x: int) -> int:
    parity = False
    while (x):
        if (x & 1):
            parity = not parity
        x >>= 1
    return parity


def collective_parity(vals: List[int]) -> int:
    collected = vals[0]
    for val in vals[1:]:
        collected ^= val
    return parity(collected)


if __name__ == '__main__':
    exit(generic_test.generic_test_main('parity.py', 'parity.tsv', parity))

File: elements-of-programming-interviews/2023/cpp/parity.cc
#include "test_framework/generic_test.h"
short Parity(unsigned long long x) {
  bool parity = false;
  int shifts = 0;
  while (x) {
    parity = (x & 1 ? !parity : parity);
    x >>= 1;
  }
  return parity ? 1 : 0;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"x"};
  return GenericTestMain(args, "parity.cc", "parity.tsv", &Parity,
                         DefaultComparator{}, param_names);
}
